x/dymns/keeper/hooks.go

```diff
@@ -0,0 +1,365 @@
+/* -------------------------------------------------------------------------- */
+
+var _ epochstypes.EpochHooks = epochHooks{}
+
+type epochHooks struct {
+	Keeper
+}
+
+// GetEpochHooks returns the epoch hooks for the module.
+func (k Keeper) GetEpochHooks() epochstypes.EpochHooks {
+	return epochHooks{
+		Keeper: k,
+	}
+}
+
+// BeforeEpochStart is the epoch start hook.
+func (e epochHooks) BeforeEpochStart(_ sdk.Context, _ string, _ int64) error {
+	return nil
+}
+
+// AfterEpochEnd is the epoch end hook.
+func (e epochHooks) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	miscParams := e.MiscParams(ctx)
+
+	if epochIdentifier != miscParams.EndEpochHookIdentifier {
+		return nil
+	}
+
+	logger := e.Logger(ctx).With("hook", "After-Epoch-End", "epoch-number", epochNumber, "epoch-identifier", epochIdentifier)
+
+	if miscParams.EnableTradingName {
+		if err := e.processActiveDymNameSellOrders(ctx, logger); err != nil {
+			return err
+		}
+	}
+
+	if miscParams.EnableTradingAlias {
+		if err := e.processActiveAliasSellOrders(ctx, logger); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// processActiveDymNameSellOrders process the finished Dym-Name Sell-Orders.
+// Sell-Order will be deleted. If the Sell-Order has a winner, the Dym-Name ownership will be transferred.
+func (e epochHooks) processActiveDymNameSellOrders(ctx sdk.Context, logger log.Logger) error {
+	finishedSOs, aSoe := e.getFinishedSellOrders(ctx, dymnstypes.TypeName, logger)
+
+	if len(finishedSOs) < 1 || len(aSoe.Records) < 1 {
+		return nil
+	}
+
+	logger.Info("processing finished SOs", "count", len(finishedSOs))
+
+	for _, so := range finishedSOs {
+		// each order should be processed in a branched context, if error, discard the state change
+		// and process next order, to prevent chain reaction when an individual order failed to process
+		errApplyStateChange := osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+			if so.HighestBid == nil {
+				e.DeleteSellOrder(ctx, so.AssetId, dymnstypes.TypeName)
+				return nil
+			}
+
+			if err := e.CompleteDymNameSellOrder(ctx, so.AssetId); err != nil {
+				logger.Error("failed to complete sell order", "asset-id", so.AssetId, "error", err)
+				return err
+			}
+
+			return nil
+		})
+
+		if errApplyStateChange == nil {
+			aSoe.Remove(so.AssetId)
+		}
+	}
+
+	if err := e.SetActiveSellOrdersExpiration(ctx, aSoe, dymnstypes.TypeName); err != nil {
+		e.Keeper.Logger(ctx).Error("failed to update active SO expiry", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// processActiveAliasSellOrders process the finished Alias Sell-Orders.
+// Sell-Order will be deleted. If the Sell-Order has a winner, the Alias linking will be updated.
+func (e epochHooks) processActiveAliasSellOrders(ctx sdk.Context, logger log.Logger) error {
+	finishedSOs, aSoe := e.getFinishedSellOrders(ctx, dymnstypes.TypeAlias, logger)
+
```

Comment:
go doc missing details about this func:
1. what about sell orders which were made for prohibited trade alias? and when can it happen?
2. we talk about finished sell orders but we also check about `aSoe` which as commented below not sure what it is (activeSellOrdersExpired? - dont know) but it's not mentioned here.

Reply:
`aSoe` is a list contains expiration date of every sell order, used to filter expired sell orders. It not a thing need to comment here.

Reply:
`and when can it happen?`
There is a big block of comment just bellow the place you written the comment:
___
			For the Sell-Orders which the assets are prohibited to trade,
			the Sell-Order will be force cancelled and the bids will be refunded.

			Why some aliases are prohibited to trade? And what are they?
			In module params, there is a list of alias mapping for some external well-known chains.
			So those aliases are considered as reserved for the external chains,
			therefor trading is not allowed.

			Why can someone own a prohibited alias?
			An alias can be bought before the reservation was made.
			But when the alias becomes reserved for the external well-known chains,
			the alias will be prohibited to trade.

			Why can someone place a Sell-Order for the prohibited alias?
			When a Sell-Order created before the reservation was made.